qiskit/optimization/algorithms/optimization_algorithm.py

```diff
@@ -104,24 +104,54 @@ class OptimizationResult:
         results: The original results object returned from the optimization algorithm. This can
             contain more information than only the optimal value and function value.
         status: The termination status of the algorithm.
+        variables: The list of variables under optimization.
     """
 
     Status = OptimizationResultStatus
 
     def __init__(self, x: Optional[Any] = None, fval: Optional[Any] = None,
                  results: Optional[Any] = None,
-                 status: OptimizationResultStatus = OptimizationResultStatus.SUCCESS) -> None:
-        self._val = x
+                 status: OptimizationResultStatus = OptimizationResultStatus.SUCCESS,
+                 variables: Optional[List[Variable]] = None) -> None:
+        self._x = x  # pylint: disable=invalid-name
+        self._variables = variables
         self._fval = fval
         self._results = results
         self._status = status
 
     def __repr__(self):
-        return '([{}] / {} / {})'.format(','.join([str(x_) for x_ in self.x]), self.fval,
-                                         self.status)
+        return 'optimal variables: [{}]\noptimal function value: {}\nstatus: {}' \
+            .format(','.join([str(x_) for x_ in self.x]), self.fval, self.status.name)
 
-    def __str__(self):
-        return 'x=[{}], fval={}'.format(','.join([str(x_) for x_ in self.x]), self.fval)
+    def __getitem__(self, item: Union[int, str]):
+        if isinstance(item, int):
+            return self.x[item]
+        if isinstance(item, str):
+            return self.variables[item]
+        raise QiskitOptimizationError("Integer or string parameter required, instead "
```

Comment:
This is confusing as in the constructor there's a definition `self._variables = ....`. Would it be better to leave `variables` as a public interface and internally work with `self._variables` only?
If variable names are not set (and they are optional now) the line `self.variables[item]` will cause an exception.

Reply:
Agree. Every time `self.variables[item]` is called, `dict(zip(self.x_names, self.x))` is invoked. It is time consuming. It would be better to construct the dictionary of variable name and value; but, it is not easy because `x` is optional by design. I would like to make it mandatory...